File: queue/Cqueue.java
public class CQueue {
    /** CQueue is the class for the implementation of circular queue*/
    private int rear, front, capacity, size;
    private int cqueue[];
    CQueue(int capacity){
        this.capacity = capacity;
        this.rear = -1;
        this.front = -1;
        this.cqueue = new int[this.capacity];
        this.size = 0;
    }
    public Boolean isEmpty(){
        /** Checks if the circular queue is empty*/
        return this.rear == -1 && this.front == -1;
    }

    public Boolean isFull(){
        /** Checks if the circular queue is full*/
        return (this.rear == this.capacity && this.front ==0)  || (this.front == this.rear +1);
    }

    public void enqueue(int item){
        /** This method performs enqueue operation on the circular queue*/
        if (this.isFull()){
            System.out.println("Queue is full. Enqueue operation cant be performed...");
            System.exit(1);
        }
        if (this.isEmpty()){
            this.front++;
        }
        this.rear = (this.rear + 1) % this.capacity;
        this.cqueue[this.rear] = item;
        this.size++;
        System.out.println("Inserted : "+ item);
    }

    public int dequeue(){
        /** This method performs dequeue operation on the circular queue*/
        if(this.isEmpty()){
            System.out.println("Queue is empty. Dequeue operation cant be performed...");
            System.exit(1);
        }
        int temp = this.cqueue[this.front];
        if (this.front == this.rear){
            this.front = -1;
            this.rear = -1;
        }
        else{
            this.front = (this.front + 1) % this.capacity;
        }
        this.size--;
        System.out.println("Removed : "+temp);
        return temp;
    }
    public int peek(){
        /** This method returns the front element of the queue*/
        if(this.isEmpty()){
            System.out.println("Queue is empty. Peek operation cant be performed...");
            System.exit(1);
        }
        return this.cqueue[this.front];

    }
    public int size(){
        /** This method returns the size of the queue*/
        return this.size;
    }
    public void display(){
        /** This method displays the elements of the queue*/
        if(this.isEmpty()){
            System.out.println("Queue is empty. Nothing is there to display");
            System.exit(1);
        }else {
            int i = this.front;
            System.out.print("CQueue : " + this.cqueue[i]);
            do {
                i = (i + 1) % this.capacity;
                System.out.print(", " + this.cqueue[i]);
            } while (i != this.rear);
            System.out.println("");
        }
    }
}

class Test {
    public static void main(String args[]) {
        CQueue cqueue = new CQueue(5);
        //System.out.println(queue.peek()); // will raise error
        //System.out.println("dequeue "+ queue.dequeue()); // will raise error
        cqueue.enqueue(-2);
        cqueue.enqueue(5);
        cqueue.display();
        System.out.println(cqueue.peek());
        System.out.println(cqueue.size());
        cqueue.enqueue(6);
        System.out.println(cqueue.size());
        cqueue.display();
        cqueue.dequeue();
        cqueue.display();
        System.out.println(cqueue.peek());
        System.out.println(cqueue.size());
        cqueue.dequeue();
        cqueue.dequeue();
        // cqueue.dequeue(); //Queue empty error
        //cqueue.display();  //Queue empty error
        //System.out.println(cqueue.peek());  //Queue empty error
        System.out.println(cqueue.size());

        cqueue.enqueue(5);
        cqueue.enqueue(10);
        cqueue.enqueue(15);
        cqueue.dequeue();
        cqueue.dequeue();
        cqueue.enqueue(20);
        cqueue.enqueue(25);
        cqueue.enqueue(30);
        cqueue.enqueue(35);

        System.out.println(cqueue.size());
        System.out.println(cqueue.peek());
        cqueue.display();

        cqueue.enqueue(99);


    }
}

File: queue/cqueue.py
class CQueue:
    def __init__(self, capacity):
        self.__rear = -1
        self.__front = -1
        self.__size = 0
        self.__capacity = capacity
        self.__cqueue = [None]*capacity
        
    def isFull(self):
        """ isFull method checks if the Circular Queue is full """
        return (self.__front == 0 and self.__rear + 1 == self.__capacity ) or (self.__front == self.__rear+1)
    
    def isEmpty(self):
        """This method checks if the  Circular Queue is empty"""
        return (self.__front == -1 and self.__rear == -1)

    def enqueue(self, item):
        """This menthod will insert elements to the Circular Queue"""
        if self.isFull():
            print("Queue is full. Enqueue operation cant be performed")
        else:
            if self.isEmpty():
                self.__front += 1
            self.__rear = (self.__rear + 1) % self.__capacity
            self.__cqueue[self.__rear] = item
            self.__size += 1
            print("Inserted : {}".format(item))
    
    def dequeue(self):
        """This method removes a element from the front of the circular queue if it is non empty"""
        if self.isEmpty():
            print("Queue is empty. Dqueue operation cant be performed")
        else:
            temp = self.__cqueue[self.__front]
            self.__size -= 1
            if self.__front == self.__rear:
                self.__front = -1
                self.__rear = -1
            else:
                self.__front = (self.__front + 1) % self.__capacity
                
            print("Removed : {}".format(temp))
            return temp
        
    def peek(self):
        """This method returns a element from the front of the circular queue if it is non empty"""
        if self.isEmpty():
            print("Queue is empty. Peek operation cant be performed")
        else:
            return self.__cqueue[self.__front]
 
                
    def size(self):
        """This menthod returns the size of the circular queue"""
        return self.__size
    
    def display(self):
        if self.isEmpty():
            print("Queue is empty. Nothing is there to display")
        else:
            if self.__front <= self.__rear:
                print('CQueue : {}'.format(', '.join([str(item) for item in self.__cqueue[self.__front : self.__rear+1]])))
            else:
                print('CQueue : {}'.format(', '.join([str(item) for item in self.__cqueue[self.__front : self.__capacity] + self.__cqueue[0 : self.__rear+1]   ])))


if __name__ == '__main__':
    cqueue = CQueue(5)
    cqueue.enqueue(-2);
    cqueue.enqueue(5);
    cqueue.display();
    print(cqueue.peek());
    print(cqueue.size());
    cqueue.enqueue(6);
    print(cqueue.size());        
    cqueue.display();
    cqueue.dequeue();
    cqueue.display();
    print(cqueue.peek());
    print(cqueue.size());
    
    cqueue.dequeue();
    cqueue.dequeue();
    cqueue.dequeue(); 
    cqueue.display();
    print(cqueue.peek());
    print(cqueue.size());
    
    cqueue.enqueue(5);
    cqueue.enqueue(10);
    cqueue.enqueue(15);
    cqueue.dequeue();
    cqueue.dequeue();
    cqueue.enqueue(20);
    cqueue.enqueue(25);
    cqueue.enqueue(30);
    cqueue.enqueue(35);
    cqueue.display();
    print(cqueue.peek());
    print(cqueue.size());
    cqueue.enqueue(99);
